scripts/js/lib/api/directMapXmlToMdx.ts

```diff
@@ -0,0 +1,127 @@
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+import type {
+  Root,
+  BlockContent,
+  Heading,
+  Text,
+  PhrasingContent,
+  List,
+  ListItem,
+  Paragraph,
+} from "mdast";
+
+// This type represents a node where all descendents have a 1:1 mapping with an mdx
+// node. For example:
+// * The XML '<title>' maps directly to MDX's `# title`. Therefore, a
+//   <title> node can be a descendent of the MdxMappableXmlNode.
+// * The XML `<parameterlist>` does NOT have a 1:1 mapping in MDX (we could map
+//   to a list, but how would we map back?). Therefore, MdxMappableXmlNodes can't
+//   have this as a descendent.
+type MdxMappableXmlNode =
+  | XmlTextNode
+  | XmlListNode
+  | { title: MdxMappableXmlNode[] }
+  | { para: MdxMappableXmlNode[] }
+  | { computeroutput: XmlTextNode[] }
+  | { verbatim: XmlTextNode[] };
+type XmlTextNode = { "#text": string };
+type XmlListNode = { itemizedlist: Array<{ listitem: MdxMappableXmlNode[] }> };
+
+/**
+ * Try to map the XML tree to a markdown AST (mdast) as closely as possible.
+ */
+export function directMapXmlToMdx(nodes: MdxMappableXmlNode[]): Root {
+  return {
+    type: "root",
+    children: nodes.flatMap((n) => xmlToBlockNodes(n, 1)),
+  };
+}
+
+/*
+ * Map an XML node (and it's children) to an array of mdast `Block` nodes
+ */
+function xmlToBlockNodes(
+  node: MdxMappableXmlNode,
+  headingDepth: Heading["depth"],
+): BlockContent[] {
+  if ("title" in node)
+    return [
+      {
+        type: "heading",
+        depth: headingDepth,
+        children: node.title.map(xmlToPhrasingNode),
+      },
+    ];
+  if ("para" in node) return xmlParagraphToBlockNode(node, headingDepth);
+  if ("verbatim" in node)
+    return [
+      {
+        type: "code",
+        lang: "C",
+        value: extractText(node.verbatim).trimEnd(),
+      },
+    ];
+  if ("itemizedlist" in node) return [xmlListToMdx(node)];
+  // If the node isn't a recognised block node, try wrapping in a paragraph and
+  // converting to a phrasing node.
+  return [{ type: "paragraph", children: [xmlToPhrasingNode(node)] }];
+}
+
+/**
+ * Map an XML node (and it's children) to an array of mdast `Phrasing` nodes
+ */
+function xmlToPhrasingNode(node: MdxMappableXmlNode): PhrasingContent | Text {
+  if ("#text" in node) return { type: "text", value: node["#text"] };
+  if ("computeroutput" in node)
+    // TODO: Handle xrefs in computeroutput nodes
+    return { type: "inlineCode", value: extractText(node.computeroutput) };
+  throw new Error(`Can't map XML node: ${JSON.stringify(node)}`);
+}
+
+function xmlParagraphToBlockNode(
+  node: { para: MdxMappableXmlNode[] },
+  headingDepth: Heading["depth"],
+): BlockContent[] {
+  // Doxygen XML sometimes nests block nodes (such as lists) in paragraphs, which
+  // markdown doesn't allow. If we encounter a When processing a block node in a
```

Comment:
Is there a missing word in the second sentence?

Reply:
Oops, yes